### PR TITLE
API branch with Variant endpoints

### PR DESCRIPTION
**Pull request by joemin**
## Summary of changes
Added a Variants endpoint to get a JSON structure for a given Variant urn. In the base case, it will return a structure that looks like:
```
    {
        'experiment': {
            ...RELEVANT_EXPERIMENT_KEYS
            'scoreset': {
                ...RELEVANT_SCORESET_KEYS
                'variants': [
                    {
                        ...RELEVANT_VARIANT_KEYS
                    },
                    ...
                ]
            }
        }
    }
```
If there is too much data (i.e. too many requested variants to return), it will instead save that exactly same formatted output and save it to a zip file which the user can download at a later time.

The user can also instead paginate by making multiple queries, specifying a limit and offset. See below for examples on that.

### Automated Tests
From inside the `app` container, run:
`python manage.py test --exclude-tag=webtest api.tests.test_views.TestVariantAPIViews`
### Manual Tests
If you have a postgres dump, this is a lot easier. Once you have everything set up, use your browser to navigate to the API browser:
`http://localhost:8000/api`
You can pick an existing variant URN and play around with the api.
#### Examples
If your urn is urn:123456abcdef#1, you could go to:
`http://localhost:8000/api/variants/urn:123456abcdef?variant_id=1`
which by default will return 20 variants from one scoreset and one experiment.
You can also try out using the `limit` and `offset` query parameters like so:
`http://localhost:8000/api/variants/urn:123456abcdef?variant_id=1&limit=10` This will limit it to 10 variants
`http://localhost:8000/api/variants/urn:123456abcdef?variant_id=1&limit=10&offset=10` This will limit it to 10 variants, but skip the first 10, which we already got above. This is basically a way to iterate through the variants by making subsequent calls.
`http://localhost:8000/api/variants/urn:123456abcdef?variant_id=1&limit=51` This is over our default max of 50 variants to return, so this will return a UUID and save a JSON response and zip it.
You can retrieve it by going to:
`http://localhost:8000/api/results/YOUR_UUID_HERE` and you'll see the binary for the zipped file. If you curl this and save it to a .zip, you will be able to unzip it, e.g.:
`curl -o result.zip http://localhost:8000/api/results/YOUR_UUID_HERE`

### CODE REVIEW

**Comment by EstelleDa** [comment]
@joemin Hi Joe, I have a question. I read your codes and ran it. I tried the command  `http://localhost:8000/api/variants/urn:123456abcdef?variant_id=1&limit=10`.
It shows me the results are # 1, # 10, # 100, # 101, # 102, # 103 etc. It's same that I changed limit=5. The results are # 1, # 10, # 100, # 101 and # 102. It looks like the IDs are treated as string not number. Is it the effect that you want? If so, then how can I get variants that IDs are from 1 to 5?

**Comment by EstelleDa** [comment]
The count_data has a different order from the original file again. For example, "count_data": {"c_1": 5, "c_2": 0, "c_4": 0, "c_8": 0, "c_16": 0, "c_32": 0, "c_64": 0, "c_0.5": 7, "c_128": 0, "c_256": 0, "c_512": 0, "ambler": 3, "c_0.25": 0, "c_1024": 0, "c_total": 12}. 
The original one is c_0.25, c_0.5, c_1, c_2, c_4, c_8, c_16, c_32, c_64, c_128, c_256, c_512, c_1024, c_total, ambler. 
The data are from the Variants table that has a wrong order.

**Comment by EstelleDa** [comment]
If I put a "-" at limit, such as `http://localhost:8000/api/variants/urn:mavedb:00000001-c-3?variant_id=4&limit=-2`, it will show me an error at the website. 
image

**Comment by afrubin** [comment]
@joemin Can you update the base branch to staging?

**Comment by afrubin** [comment]
I'm having problems getting this one running. It fails the assertion on line 83 of the docker-entrypoint.sh, because it registers 11 tasks instead of 10. If I edit the assertion to 11 tasks, the server starts up correctly and the tests run to completion.

@joemin did you add a new Celery task?

Here's the context:

```python
  try:
    inspection = celery_app.control.inspect().registered_tasks()
    if inspection is None:
      sys.stdout.write("No response. Sleeping\n")
      time.sleep(5)
    else:
      tasks = list(inspection.items())[0][1]
      assert len(tasks) == 10, "Expected 10 tasks. {} tasks were registered.".format(len(tasks))
  except Exception as e:
    raise e
```

File: api/tests/test_views.py [hunk 1]
```diff
@@ -684,3 +684,131 @@ def test_can_download_metadata(self):
...
+        variant_urn = variant.urn
+        variant_id = 1
+        variant.urn = variant_urn + f'#{variant_id}'
+        variant.save()
+        response = self.client.get(
+            f"{self.variants_base_url}/{variant_urn}?variant_id={variant_id} \
+            limit=51"
+        )
+        response_json = json.loads(response.content.decode())
+        results_uuid = response_json['results_uuid']
```
Comment (afrubin) [inline api/tests/test_views.py:811]: This threw a KeyError for me.

**Comment by afrubin** [comment]
Aside from the minor issues added here, this looks good. Very excited to get this merged!

I tested the setup with @harmatt's MaveTools branch https://github.com/VariantEffect/mavetools/tree/comment_upload_scoreset using the associated Jupyter notebooks, and everything worked out great in my local test environment.

**Comment by joemin** [comment]
> @joemin Hi Joe, I have a question. I read your codes and ran it. I tried the address `http://localhost:8000/api/variants/urn:mavedb:00000001-c-5?variant_id=707&limit=10`. It shows me the results are # 1, # 10, # 100, # 101, # 102, # 103 etc. It's same that I changed limit=5. The results are # 1, # 10, # 100, # 101 and # 102. It looks like the IDs are treated as string not number. Is it the effect that you want? If so, then how can I get variants that IDs are from 1 to 5?

Ah yeah so because of what the Variant model (/mavedb/variant/models.py) looks like, I can only easily sort the Variants by using the urn: `order_by('urn')`, which is a string. Because the `variant_id` doesn't exist in the model, I can't actually sort by that.
If we wanted to address this, I think there are two solutions:
<br>

1. **Use an expensive sort function. Basically, we would have to query for *all* matching Variant objects every time, and do something like:**

```
all _variant_objects = Objects.Variants.filter(urn__startswith=variant_urn_prefix)
all_variant_objects_ordered = []
for variant_object in variant:
  variant_id = variant_object.urn.split('#')[1]
  all_variant_objects_ordered.append((variant_id, variant_object))

all_variant_objects_ordered.sort(key=lambda y: y[0])
...
```
Which might be okay, but it does feel wrong to have to get 100s or 1000s of variants every time, especially if we only want like 10 of them or something.

<br>

2. **Change the Variant object model.**

Right now, it looks like:
```
    Attributes
    ----------
    urn : `str`, optional.
        URN identifier. Autogenerated.

    hgvs_nt : `str`, required.
        The nucleotide HGVS string belonging to the variant. Should be
        'g' if hgvs_splice is also present.

    hgvs_splice : `str`, required.
        The transcript HGVS string belonging to the variant. Should use the
        'c/n' prefix.

    hgvs_pro : `str`, required.
        The protein HGVS string belonging to the variant. Should use the 'p'
        prefix

    scoreset : `ScoreSet`, required.
        The associated scoreset of the instance.

    data : `JSONField`
        The variant's numerical data.
```
We would have to add something like
```
    variant_id : `int`, optional.
        URN identifier. Autogenerated.
```
and then run a script to migrate our database and add this field to all existing Variant objects.

Both are possible, but I think for now maybe it's okay to have it in order of the string? Let me know what you think!!

<br><br><br>
> The returning results are interesting. If there're less than 10 variants, it'll present all of the results. If there're less than 20 variants, the returned results order will be # 1, # 10, # 11, # 2, # 3 etc. If there're hundreds or thousands variants, such as `http://localhost:8000/api/variants/urn:mavedb:00000001-a-2?variant_id=2853`, it'll present me the variants are # 1, # 10, # 100, # 1000, # 1001, # 1002... # 1015. No matter how I changed the id number at address, the returning results are the same. There're total 20 returned variants.

Hi yeah sorry I just added a comment to highlight this, thanks!
Basically: in general, when querying for potentially large amounts of information, there's a default response size built in. Here, I default to the first 20 variant objects so the size of the response is not too large. The max response size is 50 variants.
- To get the first 20 and then the next 20, you can use the `offset` query parameter to skip the first n variants.
- To get _all_ variants, you can use a negative number.

Let me know what you think!

**Comment by joemin** [comment]
> The count_data has a different order from the original file again. For example, "count_data": {"c_1": 5, "c_2": 0, "c_4": 0, "c_8": 0, "c_16": 0, "c_32": 0, "c_64": 0, "c_0.5": 7, "c_128": 0, "c_256": 0, "c_512": 0, "ambler": 3, "c_0.25": 0, "c_1024": 0, "c_total": 12}. The original one is c_0.25, c_0.5, c_1, c_2, c_4, c_8, c_16, c_32, c_64, c_128, c_256, c_512, c_1024, c_total, ambler. The data are from the Variants table that has a wrong order.

So because this is returned in a JSON object, it will be treated as a dictionary. I think even if we sort it before putting in the JSON object, when it gets deserialized by the client, we won't have control over the order of the keys so I think this is okay to leave for now

**Comment by afrubin** [comment]
> > The count_data has a different order from the original file again. For example, "count_data": {"c_1": 5, "c_2": 0, "c_4": 0, "c_8": 0, "c_16": 0, "c_32": 0, "c_64": 0, "c_0.5": 7, "c_128": 0, "c_256": 0, "c_512": 0, "ambler": 3, "c_0.25": 0, "c_1024": 0, "c_total": 12}. The original one is c_0.25, c_0.5, c_1, c_2, c_4, c_8, c_16, c_32, c_64, c_128, c_256, c_512, c_1024, c_total, ambler. The data are from the Variants table that has a wrong order.
> 
> So because this is returned in a JSON object, it will be treated as a dictionary. I think even if we sort it before putting in the JSON object, when it gets deserialized by the client, we won't have control over the order of the keys so I think this is okay to leave for now

Starting with Python 3.6, dictionary ordering is preserved (based on insertion order). Not sure how this interacts with the JSON serialization, but having the columns in the "right order" is nice to have. I agree this is a lower priority, though.

**Comment by afrubin** [comment]
For the variant numbering, I think there is a third option that is a variation on option 2 (adding a new field). Right now the variant numbers are sort of arbitrary, as we don't sort the input in any sensible way. We could consider adding a new field specifically for sorting/ordering variants when using the API or generating .csv output files. 

With the changes to MAVE-HGVS and the ability to easily create sortable Python objects from the variant strings, it's possible to populate this field so that we have a defined output order that depends on the variant position and mutation.

If we create a new field specifically for ordering that isn't the `variant_id`, the ID numbers for older datasets don't have to change even if they're all out of order.

**Comment by joemin** [comment]
> For the variant numbering, I think there is a third option that is a variation on option 2 (adding a new field). Right now the variant numbers are sort of arbitrary, as we don't sort the input in any sensible way. We could consider adding a new field specifically for sorting/ordering variants when using the API or generating .csv output files.
> 
> With the changes to MAVE-HGVS and the ability to easily create sortable Python objects from the variant strings, it's possible to populate this field so that we have a defined output order that depends on the variant position and mutation.
> 
> If we create a new field specifically for ordering that isn't the `variant_id`, the ID numbers for older datasets don't have to change even if they're all out of order.

it's true. do we have a deterministic way that we'd probably order the variants? or would it be kind of custom to each family of variants?

**Comment by afrubin** [comment]
> it's true. do we have a deterministic way that we'd probably order the variants? or would it be kind of custom to each family of variants?

Two step sort on the primary variant column (hgvs_nt if present, else hgvs_pro). First a numeric sort on the position number, then a sort on the mutation (single-letter nucleotide, either single- or three-letter amino acid) within each position. The target residue is obviously the same for each position so it can be discarded.

**Comment by joemin** [comment]
> > it's true. do we have a deterministic way that we'd probably order the variants? or would it be kind of custom to each family of variants?
> 
> Two step sort on the primary variant column (hgvs_nt if present, else hgvs_pro). First a numeric sort on the position number, then a sort on the mutation (single-letter nucleotide, either single- or three-letter amino acid) within each position. The target residue is obviously the same for each position so it can be discarded.

okay yeah that makes sense to me. should I put in something like a `sort_index` or `sort_id` or `index` into the model for this PR?

this still, though, wouldn't solve @EstelleDa 's point that the user can't specify which exact variants to pull down (i.e. urn:abcdef123456(#1-#5) or something). I'm not sure what users will want, but this seems like it might be something people want. maybe we'll just have to make it clear that results are returned in our custom sort order?

**Comment by joemin** [comment]
> If I put a "-" at limit, such as `http://localhost:8000/api/variants/urn:mavedb:00000001-c-3?variant_id=4&limit=-2`, it will show me an error at the website. Symbols such as =, * and letters can be handled. Variant_id can handle negative number and symbol. image
> 
> I think this one happen because of this. I don't know why it has this error. The file is api/utilities.py image

thanks @EstelleDa !! fixed it and will push the change soon

File: api/tests/test_views.py (hunk 1 above)
Comment (EstelleDa) [reply api/tests/test_views.py:820]: Interesting. It passed on my end at first, but I ran it again and it shows me the same error.

**Comment by EstelleDa** [comment]
> > If I put a "-" at limit, such as `http://localhost:8000/api/variants/urn:mavedb:00000001-c-3?variant_id=4&limit=-2`, it will show me an error at the website. Symbols such as =, * and letters can be handled. Variant_id can handle negative number and symbol. image
> > I think this one happen because of this. I don't know why it has this error. The file is api/utilities.py image
> 
> thanks @EstelleDa !! fixed it and will push the change soon

No worries! I tried it and it works now. :D

**Comment by EstelleDa** [comment]
One more strange error. It's in api/views.py.
image

**Comment by EstelleDa** [comment]
> > > it's true. do we have a deterministic way that we'd probably order the variants? or would it be kind of custom to each family of variants?
> > 
> > 
> > Two step sort on the primary variant column (hgvs_nt if present, else hgvs_pro). First a numeric sort on the position number, then a sort on the mutation (single-letter nucleotide, either single- or three-letter amino acid) within each position. The target residue is obviously the same for each position so it can be discarded.
> 
> okay yeah that makes sense to me. should I put in something like a `sort_index` or `sort_id` or `index` into the model for this PR?
> 
> this still, though, wouldn't solve @EstelleDa 's point that the user can't specify which exact variants to pull down (i.e. urn:abcdef123456(#1-#5) or something). I'm not sure what users will want, but this seems like it might be something people want. maybe we'll just have to make it clear that results are returned in our custom sort order?

Personally, I prefer at least user can specify the exact variants to see. Maybe adding the special range such as # 20 - # 35 in the future. Otherwise, it looks like we don't need to specify the variant_id on the IP address. For example, why not just include the urn such as `http://localhost:8000/api/variants/urn:mavedb:00000001-c-1?`? `variant_id=7` lets me think that I can get a specify variant. 
It's good that only return 20 variants. I love this size.

**Comment by EstelleDa** [comment]
> > > The count_data has a different order from the original file again. For example, "count_data": {"c_1": 5, "c_2": 0, "c_4": 0, "c_8": 0, "c_16": 0, "c_32": 0, "c_64": 0, "c_0.5": 7, "c_128": 0, "c_256": 0, "c_512": 0, "ambler": 3, "c_0.25": 0, "c_1024": 0, "c_total": 12}. The original one is c_0.25, c_0.5, c_1, c_2, c_4, c_8, c_16, c_32, c_64, c_128, c_256, c_512, c_1024, c_total, ambler. The data are from the Variants table that has a wrong order.
> > 
> > 
> > So because this is returned in a JSON object, it will be treated as a dictionary. I think even if we sort it before putting in the JSON object, when it gets deserialized by the client, we won't have control over the order of the keys so I think this is okay to leave for now
> 
> Starting with Python 3.6, dictionary ordering is preserved (based on insertion order). Not sure how this interacts with the JSON serialization, but having the columns in the "right order" is nice to have. I agree this is a lower priority, though.

Yup, I agree.

File: api/tests/test_views.py (hunk 1 above)
Comment (EstelleDa) [reply api/tests/test_views.py:820]: I tried the latest version. It passes now.

**Comment by joemin** [comment]
> > > > it's true. do we have a deterministic way that we'd probably order the variants? or would it be kind of custom to each family of variants?
> > > 
> > > 
> > > Two step sort on the primary variant column (hgvs_nt if present, else hgvs_pro). First a numeric sort on the position number, then a sort on the mutation (single-letter nucleotide, either single- or three-letter amino acid) within each position. The target residue is obviously the same for each position so it can be discarded.
> > 
> > 
> > okay yeah that makes sense to me. should I put in something like a `sort_index` or `sort_id` or `index` into the model for this PR?
> > this still, though, wouldn't solve @EstelleDa 's point that the user can't specify which exact variants to pull down (i.e. urn:abcdef123456(#1-#5) or something). I'm not sure what users will want, but this seems like it might be something people want. maybe we'll just have to make it clear that results are returned in our custom sort order?
> 
> Personally, I prefer at least user can specify the exact variants to see. Maybe adding the special range such as # 20 - # 35 in the future. Otherwise, it looks like we don't need to specify the variant_id on the IP address. For example, why not just include the urn such as `http://localhost:8000/api/variants/urn:mavedb:00000001-c-1?`? `variant_id=7` lets me think that I can get a specify variant. It's good that only return 20 variants. I love this size.

I changed the implementation so the specific requested variant is always returned, and always returned first. Maybe the future iteration of this can take a list of IDs and always return those, and then if there is space left within the `limit`, it returns more related ones?

or maybe there can be an additional query parameter called `requested_variants_only` that is a boolean that can only be used **without** `limit`? that will return only the requested variant(s).

let me know what you think!

**Comment by joemin** [comment]
> One more strange error. It's in api/views.py. image

thanks! looks like this was left over from a previous PR, just fixed it

**Comment by EstelleDa** [comment]
> > > > > it's true. do we have a deterministic way that we'd probably order the variants? or would it be kind of custom to each family of variants?
> > > > 
> > > > 
> > > > Two step sort on the primary variant column (hgvs_nt if present, else hgvs_pro). First a numeric sort on the position number, then a sort on the mutation (single-letter nucleotide, either single- or three-letter amino acid) within each position. The target residue is obviously the same for each position so it can be discarded.
> > > 
> > > 
> > > okay yeah that makes sense to me. should I put in something like a `sort_index` or `sort_id` or `index` into the model for this PR?
> > > this still, though, wouldn't solve @EstelleDa 's point that the user can't specify which exact variants to pull down (i.e. urn:abcdef123456(#1-#5) or something). I'm not sure what users will want, but this seems like it might be something people want. maybe we'll just have to make it clear that results are returned in our custom sort order?
> > 
> > 
> > Personally, I prefer at least user can specify the exact variants to see. Maybe adding the special range such as # 20 - # 35 in the future. Otherwise, it looks like we don't need to specify the variant_id on the IP address. For example, why not just include the urn such as `http://localhost:8000/api/variants/urn:mavedb:00000001-c-1?`? `variant_id=7` lets me think that I can get a specify variant. It's good that only return 20 variants. I love this size.
> 
> I changed the implementation so the specific requested variant is always returned, and always returned first. Maybe the future iteration of this can take a list of IDs and always return those, and then if there is space left within the `limit`, it returns more related ones?
> 
> or maybe there can be an additional query parameter called `requested_variants_only` that is a boolean that can only be used **without** `limit`? that will return only the requested variant(s).
> 
> let me know what you think!

Thanks Joe! It works. :D 
I think it's good that we can take a list of IDs in the future. `requested_variants_only` is more clear to me, but `limit` is also enough to let me know what it can do. 
I'm wondering what we'll use these returned variants to. For example, presenting them to users such as a simple way to search variants, or letting other external developers to use our data.

**Comment by EstelleDa** [comment]
> > One more strange error. It's in api/views.py. image
> 
> thanks! looks like this was left over from a previous PR, just fixed it

No worries! hahaha~ That makes sense.

**Comment by EstelleDa** [comment]
One point share. Maybe you have discussed this already. 
I notice that, even there're less than 50 variants such as only 13 variants in total, if the command has `limit>50`, it'll still ask you to download the file to see the results.
